### PR TITLE
Support decoder text embedding model in the `examples/transformers_text.py`

### PR DESCRIPTION
**Pull request by zechengz**
Support state-of-the-art decoder text embedding model (`intfloat/e5-mistral-7b-instruct`) in the `examples/transformers_text.py`
This is currently top 1 text embedding model (outperforms Voyage AI and OpenAI's embedding models)

### CODE REVIEW

File: examples/transformers_text.py [hunk 1]
```diff
@@ -46,6 +46,20 @@
...
+# ===== product_sentiment_machine_hack ======
+# Best Val Acc: 0.9273, Best Test Acc: 0.9159
+# ======== jigsaw_unintended_bias100K =======
+# Best Val Acc: 0.9571, Best Test Acc: 0.9566
+# =============== news_channel ==============
+# Best Val Acc: 0.4837, Best Test Acc: 0.4711
+# ============ fake_job_postings2 ===========
+# Best Val Acc: 0.9851, Best Test Acc: 0.9818
+# ========== imdb_genre_prediction ==========
+# Best Val Acc: 0.9500, Best Test Acc: 0.8300
```
Comment (weihua916) [inline examples/transformers_text.py:61]: Wow this is really great :)

File: examples/transformers_text.py [hunk 2]
```diff
@@ -92,18 +107,51 @@
...
 class TextToEmbedding:
     def __init__(self, model: str, pooling: str, device: torch.device):
-        self.tokenizer = AutoTokenizer.from_pretrained(model)
-        self.model = AutoModel.from_pretrained(model).to(device)
+        self.model_name = model
         self.device = device
-        self.pooling = pooling
+        self.tokenizer = AutoTokenizer.from_pretrained(model)
+        if model == "intfloat/e5-mistral-7b-instruct":
+            self.pooling = "last"
```
Comment (weihua916) [inline examples/transformers_text.py:114]: It's a bit weird not to respect the `pooling` arg here.

**Review by weihua916 — COMMENTED** [review]
Thanks. beautiful and nice results!!

Left some questions and minor comments.

File: examples/transformers_text.py [hunk 3]
```diff
@@ -194,6 +245,22 @@ def mean_pooling(last_hidden_state: Tensor, attention_mask) -> Tensor:
     return embedding.unsqueeze(1)
 
 
+def last_pooling(last_hidden_state: Tensor, attention_mask: Tensor) -> Tensor:
+    left_padding = (attention_mask[:, -1].sum() == attention_mask.shape[0])
+    if left_padding:
+        return last_hidden_state[:, -1]
+    else:
```
[the remark this answers is not in the document]
Comment (weihua916) [reply examples/transformers_text.py:252]: and I feel `left_padding` is rarely true in most cases any way.

[the remark this answers is not in the document]
Comment (weihua916) [reply examples/transformers_text.py:252]: please ignore this comment if this `left_padding` thing is a standard thing to do in LLM community. I am just wondering.

[the remark this answers is not in the document]
Comment (zechengz) [reply examples/transformers_text.py:252]: Right for decoder architecture model (causal LLM) I think most cases we will use left padding (each token attends only attends to previous tokens, so that's why we use the last token embedding as the sentence embedding)

File: examples/transformers_text.py [hunk 4]
```diff
@@ -92,18 +107,51 @@
...
-        inputs = self.tokenizer(
-            sentences,
-            truncation=True,
-            padding="max_length",
-            return_tensors="pt",
-        )
+        if self.model_name == "intfloat/e5-mistral-7b-instruct":
+            sentences = [
+                get_detailed_instruct(
+                    "Retrieve relevant knowledge and embeddings.", sentence)
```
[the remark this answers is not in the document]
Comment (zechengz) [reply examples/transformers_text.py:127]: I think this is required for this model. But seems that this (`get_detailed_instruct`) will not influence the performance too much.

File: examples/transformers_text.py (hunk 2 above)
Comment (zechengz) [reply examples/transformers_text.py:114]: For decoder model (causal LLM), only using the last token embedding (as the sentence embedding) makes sense, so I force using "last" pooling here.

File: examples/transformers_text.py [hunk 5]
```diff
@@ -78,10 +92,11 @@
     choices=[
         "distilbert-base-uncased",
         "sentence-transformers/all-distilroberta-v1",
+        "intfloat/e5-mistral-7b-instruct",
     ],
 )
 parser.add_argument("--pooling", type=str, default="mean",
-                    choices=["mean", "cls"])
+                    choices=["mean", "cls", "last"])
```
[the remark this answers is not in the document]
Comment (zechengz) [reply examples/transformers_text.py:99]: It depends. For encoder embedding model we can use mean and cls. But for decoder embedding model only the last pooling makes sense.

File: examples/transformers_text.py [hunk 6]
```diff
@@ -92,18 +107,51 @@
...
+        self.tokenizer = AutoTokenizer.from_pretrained(model)
+        if model == "intfloat/e5-mistral-7b-instruct":
+            self.pooling = "last"
+            self.model = AutoModel.from_pretrained(
+                model,
+                torch_dtype=torch.bfloat16,
+            ).to(device)
+        else:
+            self.model = AutoModel.from_pretrained(model).to(device)
+            self.pooling = pooling
```
[the remark this answers is not in the document]
Comment (zechengz) [reply examples/transformers_text.py:121]: That's ok to me.

File: examples/transformers_text.py [hunk 7]
```diff
@@ -194,6 +245,22 @@ def mean_pooling(last_hidden_state: Tensor, attention_mask) -> Tensor:
...
+    else:
+        sequence_lengths = attention_mask.sum(dim=1) - 1
+        batch_size = last_hidden_state.shape[0]
+        return last_hidden_state[
+            torch.arange(batch_size, device=last_hidden_state.device),
+            sequence_lengths]
+
+
+def get_detailed_instruct(task_description: str, query: str) -> str:
+    return f'Instruct: {task_description}\nQuery: {query}'
```
[the remark this answers is not in the document]
Comment (zechengz) [reply examples/transformers_text.py:261]: Will move it to the code.

File: examples/transformers_text.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (weihua916) [reply examples/transformers_text.py:99]: I see. thanks for the clarification. for clarity of this example code, it'd be better to remove this argument. then it's clear which argument is used to reproduce the number

File: examples/transformers_text.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (weihua916) [reply examples/transformers_text.py:252]: Thanks. sorry, I still don't fully understand. even if `left_padding` is False, we still use the past token embedding, right?

File: examples/transformers_text.py (hunk 2 above)
Comment (weihua916) [reply examples/transformers_text.py:114]: makes sense.

File: examples/transformers_text.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (weihua916) [reply examples/transformers_text.py:252]: Maybe I am missing something. Can you please add some shape annotation and comments?

File: examples/transformers_text.py [hunk 8]
```diff
@@ -114,6 +162,9 @@ def __call__(self, sentences: List[str]) -> Tensor:
                                  mask).squeeze(1).cpu())
         elif self.pooling == "cls":
             return out.last_hidden_state[:, 0, :].detach().cpu()
+        elif self.pooling == "last":
```
Comment (weihua916) [inline examples/transformers_text.py:165]: in the line `mask = inputs["attention_mask"]`, can you add a shape annotation and describe what it stores?

Comment (zechengz) [reply examples/transformers_text.py:165]: Sure

File: examples/transformers_text.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (zechengz) [reply examples/transformers_text.py]: Oh, I see what you mean. Right, `left_padding` just makes a little bit faster. Just using the `else` one should be ok. Removed the one in the `if`

File: examples/transformers_text.py [hunk 9]
```diff
@@ -172,20 +219,15 @@ def forward(self, feat: dict[str, MultiNestedTensor]) -> Tensor:
         out = self.model(input_ids=input_ids, attention_mask=mask)
 
         # Return value has the shape [batch_size, 1, text_model_out_channels]
-        if self.pooling == "mean":
-            return mean_pooling(out.last_hidden_state, mask)
-        elif self.pooling == "cls":
-            return out.last_hidden_state[:, 0, :].unsqueeze(1)
-        else:
-            raise ValueError(f"{self.pooling} is not supported.")
+        return mean_pooling(out.last_hidden_state, mask)
```
Comment (weihua916) [inline examples/transformers_text.py:222]: Why only `mean_pooling`? Is `forward` ever used?

Comment (weihua916) [inline examples/transformers_text.py:222]: Can we add that we are using mean-based pooling in the doc-string of `TextToEmbeddingFinetune`?

File: examples/transformers_text.py [hunk 10]
```diff
@@ -91,29 +104,66 @@
...
-    def __init__(self, model: str, pooling: str, device: torch.device):
-        self.tokenizer = AutoTokenizer.from_pretrained(model)
-        self.model = AutoModel.from_pretrained(model).to(device)
+    def __init__(self, model: str, device: torch.device):
+        self.model_name = model
         self.device = device
-        self.pooling = pooling
+        self.tokenizer = AutoTokenizer.from_pretrained(model)
+        if model == "intfloat/e5-mistral-7b-instruct":
+            self.pooling = "last"
```
Comment (weihua916) [inline examples/transformers_text.py:112]: Can we add a comment why we use "last" in this case?

**Review by weihua916 — APPROVED** [review]
Thanks!!

File: examples/transformers_text.py (hunk 9 above)
Comment (zechengz) [reply examples/transformers_text.py:222]: Sure
